extensions/amp-access/0.1/amp-access.js

```diff
@@ -233,6 +256,108 @@ export class AccessService {
   }
 
   /**
+   * @private
+   */
+  scheduleView_() {
+    if (this.viewer_.isVisible()) {
+      this.reportWhenViewed_();
+    }
+    this.viewer_.onVisibilityChanged(() => {
+      if (this.viewer_.isVisible()) {
+        this.reportWhenViewed_();
+      }
+    });
+  }
+
+  /**
+   * @return {!Promise}
+   * @private
+   */
+  reportWhenViewed_() {
+    if (this.reportViewPromise_) {
+      return this.reportViewPromise_;
+    }
+    log.fine(TAG, 'start view monitoring');
+    this.reportViewPromise_ = this.whenViewed_().then(
+        this.reportViewToServer_.bind(this),
+        reason => {
+          // Ignore - view has been canceled.
+          log.fine(TAG, 'view cancelled:', reason);
+          this.reportViewPromise_ = undefined;
+          throw reason;
+        });
+    return this.reportViewPromise_;
+  }
+
+  /**
+   * The promise will be resolved when a view of this document has occurred. It
+   * will be rejected if the current impression should not be counted as a view.
+   * @return {!Promise}
+   * @private
+   */
+  whenViewed_() {
+    // Viewing kick off: document is visible.
+    const unlistenSet = [];
+
+    // Once document is visible we have a competing process for triggering
+    // condition and when the document becomes invisible again.
+    let resolve, reject;
+    const promise = new Promise((aResolve, aReject) => {
+      resolve = aResolve;
```

Comment:
All the registering below could be wrapped into this resolver to avoid the deferred pattern.